File: lib/gallerydl.ts
import { spawn } from 'child_process';
import path from 'path';
import fs from 'fs';
import { getSetting } from './db';
import { getGalleryDlPath } from './binaries';
import { getCookiePath } from './cookies';
import { buildErrorDetail } from './utils';

export interface GalleryDlFile {
  filePath: string;
  thumbnailPath: string | null;
  metadata: GalleryDlMetadata;
}

export interface GalleryDlMetadata {
  title?: string;
  description?: string;
  uploader?: string;
  author?: string;
  width?: number;
  height?: number;
  filename?: string;
  extension?: string;
  category?: string;
  subcategory?: string;
  [key: string]: unknown;
}

export async function runGalleryDl(
  url: string,
  onProgress: (filesDownloaded: number) => void,
  signal?: AbortSignal
): Promise<GalleryDlFile[]> {
  const downloadPath = getSetting('download_path') ?? '.';
  const extraArgs = getSetting('gallerydl_extra_args') ?? '';

  fs.mkdirSync(downloadPath, { recursive: true });

  const jobDir = path.join(downloadPath, `gallerydl_${Date.now()}`);
  fs.mkdirSync(jobDir, { recursive: true });

  const cookiePath = getCookiePath('gallerydl');
  const args = [
    '--write-metadata',
    '--directory', jobDir,
    ...(fs.existsSync(cookiePath) ? ['--cookies', cookiePath] : []),
    ...(extraArgs ? extraArgs.split(/\s+/).filter(Boolean) : []),
    url,
  ];

  return new Promise((resolve, reject) => {
    const proc = spawn(getGalleryDlPath(), args);

    if (signal) {
      signal.addEventListener('abort', () => proc.kill('SIGTERM'));
    }

    let filesDownloaded = 0;
    const stderrLines: string[] = [];
    const errorLines: string[] = [];

    proc.stdout.on('data', (chunk: Buffer) => {
      const text = chunk.toString();
      for (const line of text.split('\n')) {
        const trimmed = line.trim();
        if (!trimmed) continue;

        // gallery-dl prints error/warning lines starting with [error] or [warning]
        if (/^\[(error|warning)\]/i.test(trimmed)) {
          errorLines.push(trimmed);
        }

        // gallery-dl outputs downloaded file paths to stdout
        if (trimmed.startsWith('/') || trimmed.startsWith(jobDir)) {
          if (
            fs.existsSync(trimmed) &&
            !trimmed.endsWith('.json') &&
            !trimmed.endsWith('.part')
          ) {
            filesDownloaded++;
            onProgress(filesDownloaded);
          }
        }
      }
    });

    proc.stderr.on('data', (chunk: Buffer) => {
      for (const line of chunk.toString().split('\n')) {
        const trimmed = line.trim();
        if (trimmed) stderrLines.push(trimmed);
      }
    });

    proc.on('close', (code) => {
      if (code !== 0) {
        fs.rmSync(jobDir, { recursive: true, force: true });
        const detail = buildErrorDetail(errorLines, stderrLines);
        return reject(new Error(`gallery-dl exited with code ${code}${detail}`));
      }

      try {
        const results = collectGalleryDlOutput(jobDir);
        onProgress(results.length);
        resolve(results);
      } catch (err) {
        reject(err);
      }
    });

    proc.on('error', (err) => {
      reject(new Error(`Failed to spawn gallery-dl: ${err.message}`));
    });
  });
}


const IMAGE_EXTENSIONS = new Set([
  '.jpg', '.jpeg', '.png', '.gif', '.webp', '.avif',
  '.bmp', '.tiff', '.tif', '.svg',
]);

const VIDEO_EXTENSIONS = new Set([
  '.mp4', '.webm', '.mkv', '.avi', '.mov', '.flv', '.wmv',
]);

function collectGalleryDlOutput(jobDir: string): GalleryDlFile[] {
  // Collect raw media entries first
  const raw: { filePath: string; metaPath: string | null; metadata: GalleryDlMetadata }[] = [];

  function walk(dir: string) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
        continue;
      }

      const ext = path.extname(entry.name).toLowerCase();
      if (!IMAGE_EXTENSIONS.has(ext) && !VIDEO_EXTENSIONS.has(ext)) continue;

      const metaPath = fullPath + '.json';
      let metadata: GalleryDlMetadata = {};
      if (fs.existsSync(metaPath)) {
        try {
          metadata = JSON.parse(fs.readFileSync(metaPath, 'utf-8')) as GalleryDlMetadata;
        } catch {
          // keep defaults
        }
      }

      raw.push({ filePath: fullPath, metaPath: fs.existsSync(metaPath) ? metaPath : null, metadata });
    }
  }

  walk(jobDir);

  // Sort by original path for stable gallery order
  raw.sort((a, b) => a.filePath.localeCompare(b.filePath));

  return renameGalleryDlFiles(raw);
}

function renameGalleryDlFiles(
  raw: { filePath: string; metaPath: string | null; metadata: GalleryDlMetadata }[],
): GalleryDlFile[] {
  const pad = raw.length > 999 ? 4 : 3;
  return raw.map((entry, i) => {
    const index = String(i + 1).padStart(pad, '0');
    const ext = path.extname(entry.filePath).toLowerCase();
    const dir = path.dirname(entry.filePath);
    const baseName = `image_${index}${ext}`;
    const newFilePath = path.join(dir, baseName);

    fs.renameSync(entry.filePath, newFilePath);

    if (entry.metaPath) {
      fs.renameSync(entry.metaPath, path.join(dir, `${baseName}.json`));
    }

    const isImage = IMAGE_EXTENSIONS.has(ext);
    return {
      filePath: newFilePath,
      thumbnailPath: isImage ? newFilePath : null,
      metadata: entry.metadata,
    };
  });
}

File: src/gallerydl.rs
//! gallery-dl subprocess wrapper.

use crate::binaries;
use crate::config::Config;
use crate::cookies;
use crate::db::{self, Db};
use serde_json::Value;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::Arc;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::sync::watch;

#[derive(Debug, Clone)]
pub struct GalleryDlFile {
    pub file_path: PathBuf,
    pub metadata: Value,
    pub media_type: String, // video | image
}

pub async fn run_gallery_dl(
    db: &Db,
    config: &Config,
    url: &str,
    progress_tx: Option<watch::Sender<f64>>,
    cancel: tokio::sync::watch::Receiver<bool>,
    output_dir: Option<PathBuf>,
) -> anyhow::Result<Vec<GalleryDlFile>> {
    let download_path = db
        .with_conn(|c| Ok(db::get_setting(c, "download_path")))?
        .unwrap_or_else(|| config.data_dir.join("media").to_string_lossy().into());
    std::fs::create_dir_all(&download_path)?;

    let job_dir = output_dir.unwrap_or_else(|| {
        let ts = chrono::Utc::now().timestamp_millis();
        PathBuf::from(&download_path).join(format!("gallerydl_{ts}"))
    });
    std::fs::create_dir_all(&job_dir)?;

    let bin = binaries::get_gallerydl_path(db, config);
    let extra = db
        .with_conn(|c| Ok(db::get_setting(c, "gallerydl_extra_args")))?
        .unwrap_or_default();

    let mut args: Vec<String> = vec![
        "--write-metadata".into(),
        "--directory".into(),
        job_dir.to_string_lossy().into(),
    ];
    let cookies = if let Some(cookie) = cookies::cookie_path(config, "gallerydl") {
        if cookie.exists() {
            args.push("--cookies".into());
            args.push(cookie.to_string_lossy().into());
            true
        } else {
            false
        }
    } else {
        false
    };
    for part in extra.split_whitespace() {
        args.push(part.to_string());
    }
    args.push(url.to_string());

    let binary_name = Path::new(&bin)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("gallery-dl");
    tracing::info!(
        url = %url,
        binary = %binary_name,
        cookies,
        "spawning gallery-dl"
    );

    let mut child = Command::new(&bin)
        .args(&args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .spawn()?;

    let stdout = child.stdout.take().unwrap();
    let stderr = child.stderr.take().unwrap();
    let mut reader = BufReader::new(stdout).lines();
    let progress_tx = progress_tx.map(Arc::new);
    let mut files_seen = 0u32;

    let read_task = async {
        while let Ok(Some(line)) = reader.next_line().await {
            let trimmed = line.trim();
            if trimmed.starts_with('/') && Path::new(trimmed).exists() {
                let p = Path::new(trimmed);
                if p.extension()
                    .and_then(|e| e.to_str())
                    .map(|e| e != "json" && e != "part")
                    .unwrap_or(false)
                {
                    files_seen += 1;
                    if let Some(ref tx) = progress_tx {
                        let pct = (files_seen as f64 * 5.0).min(95.0);
                        let _ = tx.send(pct);
                    }
                }
            }
        }
        files_seen
    };

    let stderr_task = async {
        let mut reader = BufReader::new(stderr).lines();
        let mut tail: Vec<String> = Vec::new();
        while let Ok(Some(line)) = reader.next_line().await {
            if tail.len() >= 8 {
                tail.remove(0);
            }
            tail.push(line);
        }
        tail
    };

    let wait_task = async {
        loop {
            tokio::select! {
                status = child.wait() => break status,
                _ = async {
                    let mut c = cancel.clone();
                    let _ = c.changed().await;
                } => {
                    if *cancel.borrow() {
                        let _ = child.start_kill();
                        break child.wait().await;
                    }
                }
            }
        }
    };

    let (count, stderr_tail, status) = tokio::join!(read_task, stderr_task, wait_task);
    let status = status?;
    if !status.success() {
        if *cancel.borrow() {
            anyhow::bail!("cancelled");
        }
        let stderr_tail = truncate_stderr_tail(&stderr_tail);
        tracing::error!(
            url = %url,
            status = %status,
            stderr_tail = %stderr_tail,
            "gallery-dl exited with non-zero status"
        );
        anyhow::bail!("gallery-dl exited with {status}");
    }
    let _ = count;

    let files = collect_files(&job_dir)?;
    tracing::info!(
        url = %url,
        file_count = files.len(),
        "gallery-dl download finalized"
    );
    if let Some(tx) = progress_tx {
        let _ = tx.send(100.0);
    }
    Ok(files)
}

fn truncate_stderr_tail(lines: &[String]) -> String {
    let joined = lines.join(" | ");
    if joined.chars().count() <= 500 {
        joined
    } else {
        joined.chars().take(500).collect::<String>() + "…"
    }
}

fn collect_files(job_dir: &Path) -> anyhow::Result<Vec<GalleryDlFile>> {
    let mut media_paths: Vec<PathBuf> = Vec::new();
    for entry in walkdir::WalkDir::new(job_dir).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();
        if !path.is_file() {
            continue;
        }
        let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
        if name.ends_with(".json") || name.ends_with(".part") {
            continue;
        }
        let ext = path
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_lowercase();
        if matches!(
            ext.as_str(),
            "jpg" | "jpeg" | "png" | "gif" | "webp" | "avif" | "bmp" | "tiff"
                | "mp4" | "webm" | "mkv" | "mov"
        ) {
            media_paths.push(path.to_path_buf());
        }
    }
    media_paths.sort();

    let mut results = Vec::new();
    for (i, src) in media_paths.into_iter().enumerate() {
        let ext = src
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("jpg");
        let dest = job_dir.join(format!("image_{:03}.{ext}", i + 1));
        if src != dest {
            std::fs::rename(&src, &dest)?;
        }
        let meta_src = PathBuf::from(format!("{}.json", src.display()));
        // after rename, metadata may still use old name — also try dest.json
        let meta = read_sidecar_meta(&meta_src)
            .or_else(|| read_sidecar_meta(&PathBuf::from(format!("{}.json", dest.display()))))
            .unwrap_or(Value::Object(Default::default()));
        if meta_src.exists() {
            let _ = std::fs::rename(&meta_src, format!("{}.json", dest.display()));
        }
        let media_type = if matches!(ext, "mp4" | "webm" | "mkv" | "mov") {
            "video"
        } else {
            "image"
        };
        results.push(GalleryDlFile {
            file_path: dest,
            metadata: meta,
            media_type: media_type.into(),
        });
    }
    Ok(results)
}

fn read_sidecar_meta(path: &Path) -> Option<Value> {
    let data = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&data).ok()
}
